migrations: Add backfill to update personal DMs to rely on DM group.

mreza-kiani opened the pull request:
This migration updates recipients of 1:1 or self messages to rely on the direct message group. The affected models include `Draft`, `ScheduledMessage`, `Message`, and `ArchivedMessage`. We fetch all records and update the recipient type in batches for all messages between user1 and user2.

Fixes: backfill migration of https://github.com/zulip/zulip/issues/25713 

This PR should be merged after https://github.com/zulip/zulip/pull/35769.

**Screenshots and screen captures:**
I ran the `populate_db` multiple times to create direct messages using personal recipients.

<details>

<summary>Commands</summary>

Creating the data by removing existing data:
```bash
python manage.py populate_db   --num-messages 10000   --personals 100   --percent-personals 85   --extra-users 100 --threads 20
```

Only adding personal messages on top of existing data (needs some modification to the `populate_db` script, since it fails with the `--nodelete` flag):
```bash
python manage.py populate_db   --num-messages 10000   --personals 1000   --percent-personals 85   --nodelete --threads 20
```

</details>

I created more than 50k personal messages:
<img width="1003" height="79" alt="telegram-cloud-document-4-5843811036414613911" src="https://github.com/user-attachments/assets/13fb87d1-396a-4fd6-a084-129f46d6136b" />

I ran the migration, and it completed quickly:
![telegram-cloud-photo-size-4-5843811036870855167-x](https://github.com/user-attachments/assets/6f11dc75-da9c-4df6-bc25-4676ff0a91ad)

The message model recipient types after the migration:
![telegram-cloud-photo-size-4-5843811036870855168-x](https://github.com/user-attachments/assets/399b03c9-0867-4976-8191-a5f169490ca8)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

mreza-kiani commented:
@zulipbot add "buddy review" "mentor review"

mreza-kiani commented:
@roanster007 @sumanthvrao the migration PR is ready for review.

alexmv commented on zerver/migrations/0772_backfill_update_recipient_type_of_personal_messages_to_dm_group.py:
```diff
@@ -0,0 +1,154 @@
+                    is_user_active=is_active,
+                )
+                for user_profile_id, is_active in UserProfile.objects.filter(id__in=id_list)
+                .distinct("id")
```
This `distinct` doesn't seem necessary?

alexmv commented on zerver/migrations/0772_backfill_update_recipient_type_of_personal_messages_to_dm_group.py:
```diff
@@ -0,0 +1,154 @@
+    sender_column_name = "user_profile" if is_draft_model(message_model) else "sender"
+    personal_messages = message_model.objects.filter(
+        recipient__type=PERSONAL_RECIPIENT
+    ).prefetch_related("recipient", sender_column_name)[:BATCH_SIZE]
```
User is a very wide table, and we're only using it for its `recipient_id`, AFAICT.  This is a good place to use `.only()`.

It should also get an `order_by("id")` so it has a consistent ordering.

mreza-kiani replied:
Great point. Applied your suggestions.

mreza-kiani commented on zerver/migrations/0772_backfill_update_recipient_type_of_personal_messages_to_dm_group.py:
```diff
@@ -0,0 +1,154 @@
+            if num_of_messages > 0:
+                print(f"{message_model.__name__}: Detected {num_of_messages} personal messages.")
+                num_of_messages_updated = 0
+                while num_of_messages_updated < num_of_messages:
```
Ah, interesting. Thank you for capturing this. I skipped the initial count.

mreza-kiani commented:
I created some data:
```python
In [7]: Message.objects.values("recipient__type").annotate(count=Count('id')).order_by('recipient__type')
Out[7]: <QuerySet [{'recipient__type': 1, 'count': 25599}, {'recipient__type': 3, 'count': 14401}]>
```

And was able to run the migration successfully:

```Operations to perform:
  Target specific migration: 0779_backfill_update_recipient_type_of_personal_messages_to_dm_group, from zerver
Running migrations:
  Applying zerver.0779_backfill_update_recipient_type_of_personal_messages_to_dm_group...
Draft: Migration completed.
ScheduledMessage: Migration completed.
Message: Updated 11311 messages in this batch.
Message: Updated 8472 messages in this batch.
Message: Updated 5816 messages in this batch.
Message: Migration completed.
ArchivedMessage: Migration completed.
 OK
```

Messages after the migration:
```python
In [8]: Message.objects.values("recipient__type").annotate(count=Count('id')).order_by('recipient__type')
Out[8]: <QuerySet [{'recipient__type': 3, 'count': 40000}]>
```

alexmv commented on zerver/migrations/0779_backfill_update_recipient_type_of_personal_messages_to_dm_group.py:
```diff
@@ -0,0 +1,166 @@
+                     wildcard_mentions_notify)
+                SELECT %s, id, is_active,
+                       true, false, '#c2c2c2', false,
+                       NULL, NULL, NULL, NULL, NULL
```
All of the nullable columns can be left off of this; they default to NULL.

Will this choice of `color` come back to bite us?  It's currently unused, right?

alexmv reviewed (commented):
This is looking good.

One thing that I'd like to make sure we've verified is what happens when you take your test instance, and _half-run_ the migration.  That is, I want to make sure we've at least briefly hand-tested the state where some of the recipients have moved, but not all.  Double-check to make sure that message sending and fetching seems to run as expected in that case -- this is particularly important, because that state will be true in Zulip Cloud for a _while_, and if something is broken in that state, we'll have no recourse except to just ... wait until the migration completes.  Which would be awkward.

alexmv commented on zerver/migrations/0788_backfill_update_recipient_type_of_personal_messages_to_dm_group.py:
```diff
@@ -0,0 +1,169 @@
+                        [personal_rec_id],
+                    )
+                    for (sender_id,) in cursor.fetchall():
+                        sender_tables.setdefault(sender_id, []).append((table, col))
```
We want the information about which tables have which sender-ids -- that saves a lot more round-trips in the UPDATE than the 3 queries we'd save in the initial setup.

We could do it as one `UNION` (the `ALL` would be a noop here)  with each query having its own tablename as a column.
